stdpopsim/annotations.py

```diff
@@ -0,0 +1,133 @@
+"""
+Infrastructure for defining information about genome annotation.
+"""
+import logging
+import pybedtools
+import attr
+
+import pathlib
+import contextlib
+import warnings
+import os
+import urllib.request
+from . import cache
+
+logger = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def cd(path):
+    """
+    Convenience function to change the current working directory in a context
+    manager.
+    """
+    old_dir = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(old_dir)
+
+
+@attr.s
+class GTF(object):
+    """
+    Class represnting a GTF file that we will
+
+    :ivar url: The URL where the packed and compressed GTF can be found
+    :vartype url: str
+    :ivar id: species id
+    :vartype id: str
+    :ivar species: a `stdpopsim.species` instance
+    :ivar description: description of annotation file
+    :vartype description: str
+    """
+    url = attr.ib(default=None)
+    species = attr.ib(default=None)
+    id = attr.ib(default=None)
+    file_name = attr.ib(default=None)
+    description = attr.ib(default=None)
+    citations = attr.ib(factory=list)
+    long_description = attr.ib(default=None)
+
+    def __attrs_post_init__(self):
+        self.file_name = os.path.basename(self.url)
+
+    @property
+    def cache_dir(self):
+        return pathlib.Path(cache.get_cache_dir()) / "annotations"
+
+    @property
+    def species_cache_dir(self):
+        return self.cache_dir / self.species.id
+
+    def __str__(self):
+        s = "GTF Annotation:\n"
+        s += "\tspecies   = {}\n".format(self.species.name)
+        s += "\tid        = {}\n".format(self.id)
+        s += "\turl       = {}\n".format(self.url)
+        s += "\tcached    = {}\n".format(self.is_cached())
+        s += "\tcache_dir = {}\n".format(self.species_cache_dir)
+        return s
+
+    def is_cached(self):
+        """
+        Returns True if this map is cached locally.
+        """
+        return os.path.exists(self.species_cache_dir)
+
+    def download(self):
```

Comment:
It might be nice to factor out the duplicated download/cache code between this class and the GeneticMap class.

Reply:
the directory structure here is different enough that it would be a pain i think. little reward vs the bit of duplicated code IMO